File: Python/Algorithms/Sieve Algorithms/Sieve of Sundaram.py
"""
    Sieve of Sundaram :
    Given a number n, print all primes smaller than or equal to n.
"""


def sieve_of_sundaram(n):
    # For primes smaller than n reduce n to half
    n_cap = int((n - 1) / 2)
    # Make all elements unmarked
    marked = [0] * (n_cap + 1)

    # Mark all numbers of the form i + j + 2ij
    for i in range(1, n_cap + 1):
        j = i
        while((i + j + 2 * i * j) <= n_cap):
            marked[i + j + 2 * i * j] = 1
            j += 1
    primes = []

    # 2 is a prime
    if (n > 2):
        primes.append(2)

    # Numbers of the form 2 * i + 1 are prime
    for i in range(1, n_cap + 1):
        if (marked[i] == 0):
            primes.append(2 * i + 1)

    return primes


num = int(input("Please enter the number: "))
# print(sieve_of_sundaram(25))
print(sieve_of_sundaram(num))

File: C++/Algorithms/Sieve Algorithms/Sieve_of_Sundaram.cpp
/*
    Algorithm: Sieve of Sundaram. 
    Given a number n, print all prime numbers less than n
*/

#include <bits/stdc++.h> 
using namespace std; 
  
int sieve_of_sundaram(int n) 
{ 
    // Sieve of Sundaram produces primes smaller than (2*x + 2) for a number given number x. 
    int m = (n-1)/2; 
  
    // This array is used to identify prime number
    vector<bool> visited(m + 1, false);

    //Core : Mark all numbers of the form i + j + 2ij as true where 1 <= i <= j 
    for (int i=1; i<=m; i++) 
        for (int j=i; (i + j + 2*i*j) <= m; j++) 
            visited[i + j + 2*i*j] = true; 
  
    // Since 2 is a prime number, and we will be printing using 2*i + 1 , so 2 will be skipped in this case.
    if (n > 2) cout << 2 << " "; 
    
    // Printing Primes
    for (int i=1; i<=m; i++)
        if (!visited[i])
            cout << 2*i + 1 << " "; 
} 
  
int main(void) 
{ 
    sieve_of_sundaram(50); 
    return 0; 
} 